ping/ping_common.c

```diff
@@ -430,8 +430,13 @@ int pinger(struct ping_rts *rts, ping_func_set_st *fset, socket_st *sock)
 	if (i == 0 && !rts->opt_quiet) {
 		if (rts->opt_flood)
 			write_stdout("E", 1);
-		else
-			error(0, errno, "sendmsg");
+		else {
+			if(!rts->opt_json)
+				error(0, errno, "sendmsg");
+
+			error_json(rts, 0, "sendmsg", strerror(errno), PING_JSON_NUL, "");
```

Comment:
As I wrote I'm not sure if error msg is really important to have in JSON. But if you prepared it, it'd be great to be  `if (!rts->opt_json)` was handled internally. I.e. have custom macro / inline function, which either call `error()` or `error_json()` based on `if (!rts->opt_json)`.

Reply:
I also implemented a macro for this now, but it is not so useful yet.
It's cumbersome to facilitate all cases in a macro or function, because the arguments passed to `error()` in the current code vary a lot. Either the message is only a type of error, like "sendmsg" (ok for JSON error type value). Or it is a long error message, like "Message too long" (ok for JSON error message value). Or it is a warning message, like "WARNING: probably, rcvbuf is not enough to hold preload" (does not fit in JSON error map at all).
In some places it is also a format string, which is particularly problematic to implement in the macro because I also use variadic arguments `error_json()`.

However I might consider dropping the latter - the additional nesting in the JSON to show different "types" of errors which I currently have might be over engineering, and I should simply make a key "error" and a value containing whatever message would otherwise be printed using `error()`.

Reply:
Sounds good, I like `ERROR()`.

BTW I know it's a draft and you probably plan it, but final version should not have commits which was later removed, e.g. libjson added and then removed etc (you can fix it with `git rebase -i master` and `git push -f`).

Reply:
Sure, as already mentioned in https://github.com/iputils/iputils/pull/571#issuecomment-2518355436 I will clean up commits before marking as ready.